### PR TITLE
snapshot/cli: cli to snapshot db

### PR DESCRIPTION
**Pull request by dbentley**
scoot-snapshot-db create ingest_git_commit 545c88d71d40a49ebdfb1d268c724110793330d2

gives stdout of:

local-gc-545c88d71d40a49ebdfb1d268c724110793330d2

(don't yet have uploading turned on in this branch)

### CODE REVIEW

File: snapshot/cli/cli.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	createCmd := &cobra.Command{
+		Use:   "create",
+		Short: "create a snapshot",
+	}
+	rootCmd.AddCommand(createCmd)
+	parentCmd = createCmd
+
+	add(&ingestGitCommitCommand{})
```
Comment (jasonbs10) [inline snapshot/cli/cli.go]: Add() should probably take cobra parentCmd as a param and return the new cobra cmd.

File: snapshot/cli/cli.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func MakeDBCLI(injector DBInjector) *cobra.Command {
+	rootCmd := &cobra.Command{
+		Use:   "scoot-snapshot-db",
+		Short: "scoot snapshot db CLI",
+	}
+	parentCmd := rootCmd
+	add := func(subCmd dbCommand) {
+		cmd := subCmd.register()
```
Comment (jasonbs10) [inline snapshot/cli/cli.go]: Kind of confusing to switch between dbCommand and cobra.Command without tagging the vars as such.

File: binaries/scoot-snapshot-db/main.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/spf13/cobra"
+
+	"github.com/scootdev/scoot/os/temp"
+	"github.com/scootdev/scoot/snapshot"
+	"github.com/scootdev/scoot/snapshot/cli"
+	"github.com/scootdev/scoot/snapshot/git/gitdb"
+	"github.com/scootdev/scoot/snapshot/git/repo"
+)
+
+func main() {
```
Comment (jasonbs10) [inline binaries/scoot-snapshot-db/main.go]: There should be a help string stating that cwd must be a the git repo (even if it's implied since there's no other way to specify it)

File: binaries/scoot-snapshot-db/main.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	inj := &injector{}
+	cmd := cli.MakeDBCLI(inj)
+	if err := cmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type injector struct{}
+
+func (i *injector) RegisterFlags(rootCmd *cobra.Command) {}
```
Comment (jasonbs10) [inline binaries/scoot-snapshot-db/main.go]: I believe this is an implementation review and if that's the case, we should remove things that aren't obviously needed (or add a TODO if you wish).

File: snapshot/cli/cli.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	cmd := &cobra.Command{
+		Use:   "ingest_git_commit",
+		Short: "ingest a git commit",
+	}
+	cmd.Flags().StringVar(&c.commit, "commit", "", "commit to ingest")
+	return cmd
+}
+
+func (c *ingestGitCommitCommand) run(db snapshot.DB, _ *cobra.Command, _ []string) error {
+	wd, err := os.Getwd()
```
Comment (jasonbs10) [inline snapshot/cli/cli.go]: Might be better to construct this command with a repoDir.

File: binaries/scoot-snapshot-db/main.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+func (i *injector) Inject() (snapshot.DB, error) {
```
Comment (jasonbs10) [inline binaries/scoot-snapshot-db/main.go]: Why do we need an injector? We may want to delay db construction but we can't we instead pass the necessary params into MakeDBCLI()?

File: binaries/scoot-snapshot-db/main.go (hunk 4 above)
Comment (dbentley) [inline binaries/scoot-snapshot-db/main.go]: Done.

File: binaries/scoot-snapshot-db/main.go (hunk 6 above)
Comment (dbentley) [inline binaries/scoot-snapshot-db/main.go]: So, I maybe buried the lede:

In all our other CLI tools, we've had a problem trying to mix args. So we've had to setup first, then call the CLI, and trying to combine has gotten messy (especially with the --addr flags).

This is nifty because it allows us to do everything in cobra: the injector can register flags (though doesn't right now), and then those get parsed naturally by cobra.

So, at some point (and soon), we're going to want the injector to have args, and it's going to be able to; yay!

To summarize:
the injector doesn't just delay db construction; it allows getting args from the CLI, using them to make objects, and *then* calling into the CLI's shared logic, across our open-source and closed-source binaries.

File: binaries/scoot-snapshot-db/main.go (hunk 3 above)
Comment (dbentley) [inline binaries/scoot-snapshot-db/main.go]: Hrm, not sure exactly how to it that in...

I added:

	dataRepo, err := repo.NewRepository(wd)
	if err != nil {
		return nil, fmt.Errorf(
			"cannot create a repo in wd %v; scoot-snapshot-db must be run in a git repo", wd, err)
	}


WDYT?

File: snapshot/cli/cli.go (hunk 5 above)
Comment (dbentley) [inline snapshot/cli/cli.go]: OK, I started writing a github PR comment about when what happens, and realized many of your comments were because this is poorly explained. I just left a big godoc package comment on the top of cli.go; why don't you take a look and we can go from there?

File: snapshot/cli/cli.go (hunk 1 above)
Comment (dbentley) [inline snapshot/cli/cli.go]: OK, take parentCmd as a param; not sure we need the returned cobra cmd (will we ever have a non-leaf cmd that also have behavior?)

File: snapshot/cli/cli.go (hunk 2 above)
Comment (dbentley) [inline snapshot/cli/cli.go]: Yup... Any idea how best to tag them?

**Comment by dbentley** [comment]
OK, I've updated with a lengthy doc comment and a few other things.

I am proud of what good things there are, that I've explained much more in the lengthy doc comment.

A few of your comments I agree with and am not sure how best to fix. I'm planning to merge and hope we can improve it together...

File: snapshot/cli/cli.go (hunk 2 above)
Comment (jasonbs10) [inline snapshot/cli/cli.go]: Giving all the cobra commands a cobra postfix isn't too horrible.

File: binaries/scoot-snapshot-db/main.go (hunk 3 above)
Comment (jasonbs10) [inline binaries/scoot-snapshot-db/main.go]: That should be fine actually.
